### PR TITLE
BIGTOP-2345 Create Flink packaging

### PR DESCRIPTION
**Pull request by rmetzger**
This PR is based on the work done in https://github.com/apache/bigtop/pull/93.
I've picked only the parts regarding the packaging (ignoring the puppet deployment).

The changes have been tested on a Debian8 virtual machine and Fedora 23.
Packaging, installation, the init scripts, user creation, file removal, logging seem to work now.

I'm looking forward to any comments regarding the proposed changes.

### CODE REVIEW

File: bigtop-packages/src/rpm/flink/SPECS/flink.spec [hunk 1]
```diff
@@ -0,0 +1,143 @@
...
+
+###### FILES ###########
+
+%files 
+%defattr(-,root,root,755)
+%config(noreplace) %{config_flink}.dist
+
+%dir %{_sysconfdir}/%{flink_name}
+%config(noreplace) %{initd_dir}/%{flink_name}-master
+%config(noreplace) %{initd_dir}/%{flink_name}-worker
```
Comment (mxm) [inline bigtop-packages/src/rpm/flink/SPECS/flink.spec]: I think those need to be 
````
%config(noreplace) %{initd_dir}/%{flink_name}-jobmanager
%config(noreplace) %{initd_dir}/%{flink_name}-taskmanager
```

**Comment by mxm** [comment]
I'm getting the following error message when building using `gradle flink-rpm`:


>RPM build errors:
    File not found: /home/max/Desktop/bigtop/build/flink/rpm/BUILDROOT/flink-1.0.0-1.fc23.x86_64/etc/rc.d/init.d/flink-master
    File not found: /home/max/Desktop/bigtop/build/flink/rpm/BUILDROOT/flink-1.0.0-1.fc23.x86_64/etc/rc.d/init.d/flink-worker
    File not found: /home/max/Desktop/bigtop/build/flink/rpm/BUILDROOT/flink-1.0.0-1.fc23.x86_64/usr/share/doc/flink-1.0.0

File: bigtop-packages/src/rpm/flink/SPECS/flink.spec [hunk 2]
```diff
@@ -0,0 +1,145 @@
...
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+%define flink_name flink
+%define lib_flink /usr/lib/%{flink_name}
+%define bin_flink /usr/bin
+%define etc_flink /etc/%{flink_name}
+%define config_flink %{etc_flink}/conf
+%define man_dir %{_mandir}
+%define flink_services master worker
```
Comment (bhupendrasingh) [inline bigtop-packages/src/rpm/flink/SPECS/flink.spec]: I think this also needs to be changed to -
%define flink_services jobmanager taskmanager

File: bigtop-packages/src/rpm/flink/SPECS/flink.spec (hunk 1 above)
Comment (rmetzger) [inline bigtop-packages/src/rpm/flink/SPECS/flink.spec]: It seems that the rebasing before the PR went horribly wrong :(

**Comment by rmetzger** [comment]
The gradle build of the RPM should be fixed. But there seems to be still an issue with the systemd start of the jobmanager.

**Comment by rmetzger** [comment]
I verified the RPM build again. It should be working now.
I'll now test the debian packaging again.

**Comment by rmetzger** [comment]
Debian is now fixed as well.

**Comment by mxm** [comment]
Tested the RPM package on Fedora 23. First, a bit of a disappointment: 

```bash
● jobmanager.service - LSB: Flink jobmanager
   Loaded: loaded (/etc/rc.d/init.d/jobmanager)
   Active: failed (Result: exit-code) since Mon 2016-04-25 16:40:45 CEST; 16s ago
     Docs: man:systemd-sysv-generator(8)
  Process: 13344 ExecStart=/etc/rc.d/init.d/jobmanager start (code=exited, status=127)

Apr 25 16:40:41 localhost.localdomain jobmanager[13344]: /etc/rc.d/init.d/jobmanager: line 34: /lib/lsb/init-functions: No such file or directory
Apr 25 16:40:41 localhost.localdomain jobmanager[13344]: /etc/rc.d/init.d/jobmanager: line 74: log_success_msg: command not found
Apr 25 16:40:41 localhost.localdomain jobmanager[13344]: /etc/rc.d/init.d/jobmanager: line 120: pidofproc: command not found
Apr 25 16:40:41 localhost.localdomain su[13386]: (to flink) root on none
Apr 25 16:40:45 localhost.localdomain jobmanager[13344]: /etc/rc.d/init.d/jobmanager: line 120: pidofproc: command not found
Apr 25 16:40:45 localhost.localdomain jobmanager[13344]: /etc/rc.d/init.d/jobmanager: line 102: log_failure_msg: command not found
Apr 25 16:40:45 localhost.localdomain systemd[1]: jobmanager.service: Control process exited, code=exited status=127
Apr 25 16:40:45 localhost.localdomain systemd[1]: Failed to start LSB: Flink jobmanager.
Apr 25 16:40:45 localhost.localdomain systemd[1]: jobmanager.service: Unit entered failed state.
Apr 25 16:40:45 localhost.localdomain systemd[1]: jobmanager.service: Failed with result 'exit-code'.
```

Turns out, it works when you install `lsb` or more specifically `redhat-lsb-core`:

```bash
● jobmanager.service - LSB: Flink jobmanager
   Loaded: loaded (/etc/rc.d/init.d/jobmanager)
   Active: active (exited) since Mon 2016-04-25 16:42:56 CEST; 2s ago
     Docs: man:systemd-sysv-generator(8)
  Process: 14071 ExecStart=/etc/rc.d/init.d/jobmanager start (code=exited, status=0/SUCCESS)

Apr 25 16:42:56 localhost.localdomain systemd[1]: Starting LSB: Flink jobmanager...
Apr 25 16:42:56 localhost.localdomain jobmanager[14071]: Starting Flink jobmanager (flink-jobmanager):[  OK  ]
Apr 25 16:42:56 localhost.localdomain jobmanager[14071]: Flink jobmanager is running[  OK  ]
Apr 25 16:42:56 localhost.localdomain systemd[1]: Started LSB: Flink jobmanager.

● taskmanager.service - LSB: Flink taskmanager
   Loaded: loaded (/etc/rc.d/init.d/taskmanager)
   Active: active (exited) since Mon 2016-04-25 16:44:28 CEST; 22s ago
     Docs: man:systemd-sysv-generator(8)
  Process: 15428 ExecStart=/etc/rc.d/init.d/taskmanager start (code=exited, status=0/SUCCESS)

Apr 25 16:44:25 localhost.localdomain systemd[1]: Starting LSB: Flink taskmanager...
Apr 25 16:44:25 localhost.localdomain taskmanager[15428]: Starting Flink taskmanager (flink-taskmanager):[  OK  ]
Apr 25 16:44:25 localhost.localdomain su[15483]: (to flink) root on none
Apr 25 16:44:28 localhost.localdomain taskmanager[15428]: Started Flink taskmanager (flink-taskmanager):[  OK  ]
Apr 25 16:44:28 localhost.localdomain systemd[1]: Started LSB: Flink taskmanager.
```

So we need to add `lsb` as a dependency for Flink. Or do we assume people have it installed?

The jobmanager/taskmanager came up fine. I wonder, should we rename the service to `flink-master` and `flink-worker`? Or `flink-jobmanager` and `flink-taskmanager`? Simply TaskManager and JobManager could confuse people.

Another thing, the logs are named "flink-flink-jobmanager-0-localhost.localdomain.log". Probably could remove the `flink-flink` prefix.

Other than that, great work :+1:

**Comment by oflebbe** [comment]
Hi. My Nitty picky comment:

Please see https://cwiki.apache.org/confluence/display/BIGTOP/Bigtop+Packaging on how to patch sources. (you may have a look at the hive or zookeeper packages for examples)

Short : rename 1837.patch to patch1-1837.patch . Insert the magic patterns into the spec file. remove the patch command from the do-component-build.

In order to get committed the patch on the branch has to be collapsed to only one patch with the exact message as the JIRA.

**Comment by oflebbe** [comment]
Hi the build looks very promising! You should not package /var/run, since this is only a temporary filesystem. You should create it with your init script (This will fix one of the lintian errors for debian/ubuntu).

**Comment by rmetzger** [comment]
Thank you for the feedback!
I'm trying to address them soon.

**Comment by rmetzger** [comment]
Thank you for the review @mxm!

Regarding the lsb dependency: I saw that the Hadoop spec for example has the following entry:
```
Requires: sh-utils, /lib/lsb/init-functions
```
I guess we can add that as well.


> I wonder, should we rename the service to flink-master and flink-worker? Or flink-jobmanager and flink-taskmanager? Simply TaskManager and JobManager could confuse people.

I'm definitively against `flink-master`, `flink-worker`, because that's not an "official" terminology (at least in the Flink code we call them job- and taskmanager).
I also thought about `flink-jobmanager` and `flink-taskmanager`. Any other thoughts on this?

> Another thing, the logs are named "flink-flink-jobmanager-0-localhost.localdomain.log". Probably could remove the flink-flink prefix.

The naming schema for logs is `flink-<username>-<component>-<instance>-<host>.<log|out>`. By starting the flink services with the `flink` user, we end up with that ugliness.
I'll look into it and try to see if I can change the behavior easily (the problem is that I need to create patches if I want to change anything on the flink source, and I want to avoid patching too much for bigtop).

@oflebbe: I'll check the guide again and adopt the patch integration accordingly.

**Comment by rmetzger** [comment]
@oflebbe quick question:
The spec file contains already the following section:

```spec
%if  %{?suse_version:1}0
# Required for init scripts
Requires: insserv
%global initd_dir %{_sysconfdir}/rc.d

%else
# Required for init scripts
Requires: /lib/lsb/init-functions
%global initd_dir %{_sysconfdir}/rc.d/init.d
%endif
```

I assume the if condition is true for suse, so the else branch applied for Max' build, requiring the init-functions?

**Comment by rmetzger** [comment]
Another follow up Q:

> You should not package /var/run, since this is only a temporary filesystem. You should create it with your init script (This will fix one of the lintian errors for debian/ubuntu).

Okay, makes sense. I was looking a bit into other scripts and I saw that others are setting the `$WORKING_DIR` variable.
I wonder why `/bigtop-packages/src/templates/init.d.tmpl` is not taking care of the creation and deletion of the `$WORKING_DIR` for that?

**Comment by mxm** [comment]
>I'm definitively against flink-master, flink-worker, because that's not an "official" terminology (at least in the Flink code we call them job- and taskmanager).

Then let's call them `flink-jobmanager` and `flink-taskmanager`. The `flink-` prefix will help people to discover the service names.

**Comment by rmetzger** [comment]
Okay, I'll rename the services

**Comment by rmetzger** [comment]
I addressed all remaining issues with the pull request.

Please let me know if there's anything else blocking this from being merged.

**Comment by mbalassi** [comment]
Hey guys, any update on this? I am looking forward to having the Flink integration in. :)

**Comment by c0s** [comment]
I have committed this to the master. Feel free to close the PR and thanks for your contribution!

**Comment by rmetzger** [comment]
Thanks a lot for merging.
I'm closing the PR now.

**Pull request closed without merging**